rust/relay/src/main.rs

```diff
@@ -51,110 +38,159 @@ async fn main() -> Result<()> {
     Ok(())
 }
 
-struct Eventloop<'a> {
+struct Eventloop {
     ip4_socket: UdpSocket,
     listen_ip4_address: Ipv4Addr,
     server: Server,
-    allocations: HashMap<AllocationId, task::JoinHandle<()>>,
+    allocations: HashMap<AllocationId, Allocation>,
     relay_data_sender: mpsc::Sender<(Vec<u8>, SocketAddr, AllocationId)>,
     relay_data_receiver: mpsc::Receiver<(Vec<u8>, SocketAddr, AllocationId)>,
     sleep: Sleep,
-    recv_buf: ReadBuf<'a>,
+
+    client_send_buffer: VecDeque<(Vec<u8>, SocketAddr)>,
+    allocation_send_buffer: VecDeque<(Vec<u8>, SocketAddr, AllocationId)>,
 }
 
-impl<'a> Eventloop<'a> {
-    async fn new(
-        public_ip4_address: Ipv4Addr,
-        listen_ip4_address: Ipv4Addr,
-        recv_buf: ReadBuf<'a>,
-    ) -> io::Result<Eventloop<'a>> {
+struct Allocation {
+    /// The handle to the task that is running the allocation.
+    ///
+    /// Stored here to make resource-cleanup easy.
+    handle: task::JoinHandle<()>,
+    sender: mpsc::Sender<(Vec<u8>, SocketAddr)>,
+}
+
+impl Eventloop {
+    async fn new(public_ip4_address: Ipv4Addr, listen_ip4_address: Ipv4Addr) -> Result<Self> {
         let (sender, receiver) = mpsc::channel(1);
 
         Ok(Self {
             ip4_socket: UdpSocket::bind((listen_ip4_address, 3478)).await?,
             listen_ip4_address,
-            server: Server::new(SocketAddrV4::new(public_ip4_address, 3478)),
+            server: Server::new(public_ip4_address),
             allocations: Default::default(),
             relay_data_sender: sender,
             relay_data_receiver: receiver,
             sleep: Sleep::default(),
-            recv_buf,
+            client_send_buffer: Default::default(),
+            allocation_send_buffer: Default::default(),
         })
     }
 
     fn poll(&mut self, cx: &mut std::task::Context<'_>) -> Poll<Result<()>> {
         loop {
             // Priority 1: Execute the pending commands of the server.
```

Comment:
I am not sure if what I am doing is the right strategy here because we don't have backpressure in UDP anyway but the goal was to not have our internal buffers grow unbounded. We potentially have many UDP sockets (one per allocation) and one main socket. If I just take data from the allocation sockets whenever it is ready, then we may get unbounded buffer growth if our main socket cannot keep up with writing.

Without a concept of backpressure, we have to drop packets somewhere. In this implementation, that will happen likely in the network driver somewhere.

Reply:
Hm I see what you mean. We can leave it like this and revisit this when we implement bandwidth, I still don't think we should let one of the branches "freeze" the event-loop.

Perhaps we can just use a ring buffer and drop received packets ourselves if a situation like that arises.

Reply:
> I still don't think we should let one of the branches "freeze" the event-loop.

The general pattern has been working really well for us in `rust-libp2p` but there we (aim to) have back-pressure between all the components. Maybe that is not a good idea in UDP-based systems?

> Perhaps we can just use a ring buffer and drop received packets ourselves if a situation like that arises.

Before we implement something like that, I would like to instrument the code with metrics and see how often that actually happens.